File: app/pricing/selling_price.py
"""Selling-price generation via reverse unit economics (spec §8–§10)."""

from app.pricing.assumptions import DEFAULT_MARGINS, STRESS_AD_SPEND_PCT, ladder_for
from app.pricing.models import (
    Brand,
    FixedVariableCosts,
    MarginAssumptions,
    SellingPriceResult,
)


def remaining_margin(margins: MarginAssumptions, ad_spend_pct: float | None = None) -> float:
    """1 − ad − team − overhead − target profit (spec §9)."""
    ad = margins.ad_spend_pct if ad_spend_pct is None else ad_spend_pct
    return 1 - ad - margins.team_pct - margins.overhead_pct - margins.target_profit_pct


def snap_up_to_ladder(raw_sp: float, ladder: list[int]) -> int | None:
    """Smallest approved ladder price at or above the raw price (spec §8)."""
    for price in ladder:
        if price >= raw_sp:
            return price
    return None


def generate_selling_price(
    design_cp: float,
    fixed_costs: FixedVariableCosts,
    brand: Brand,
    margins: MarginAssumptions = DEFAULT_MARGINS,
) -> SellingPriceResult:
    """Recommend a ladder price that clears unit economics (spec §9–§10).

    SP = fixed variable costs / remaining margin, snapped up to the approved
    ladder. Also reports whether the price survives the 40% ad-spend stress
    test, and — for each cheaper rung it can't support — the max Design CP
    that *would* make that rung viable.
    """
    margin_normal = remaining_margin(margins)
    if margin_normal <= 0:
        raise ValueError("Remaining margin must be positive; check the margin assumptions.")

    other_fixed = fixed_costs.total_excluding_cp()
    fixed_variable = round(design_cp + other_fixed, 2)
    raw_sp = round(fixed_variable / margin_normal, 2)

    ladder = ladder_for(brand)
    recommended = snap_up_to_ladder(raw_sp, ladder)
    cp_pct_at_recommended = round(design_cp / recommended, 4) if recommended else None

    margin_stress = remaining_margin(margins, ad_spend_pct=STRESS_AD_SPEND_PCT)
    survives_stress = bool(
        recommended and margin_stress > 0 and recommended >= fixed_variable / margin_stress
    )

    warnings = _rung_warnings(design_cp, other_fixed, margin_normal, ladder, recommended)
    if recommended is None:
        warnings.append("Raw price exceeds the top of the ladder — review costs or category.")

    return SellingPriceResult(
        raw_sp=raw_sp,
        recommended_sp=recommended,
        cp_pct_at_recommended=cp_pct_at_recommended,
        passes_normal=recommended is not None,
        survives_stress=survives_stress,
        warnings=warnings,
    )


def _rung_warnings(
    design_cp: float,
    other_fixed: float,
    margin_normal: float,
    ladder: list[int],
    recommended: int | None,
) -> list[str]:
    """For each rung below the recommendation, the max Design CP that fits it."""
    warnings: list[str] = []
    for price in ladder:
        if recommended is not None and price >= recommended:
            break
        max_cp = round(price * margin_normal - other_fixed, 2)
        if max_cp >= 0:
            warnings.append(
                f"₹{price} needs Design CP ≤ ₹{max_cp:.0f} (currently ₹{design_cp:.0f})."
            )
        else:
            warnings.append(f"₹{price} is not viable at these fixed costs.")
    return warnings

File: internal/pricing/selling_price.go
package pricing

import "fmt"

// RemainingMargin is the share of the selling price left after the margin
// components. If adSpendPct is non-nil it overrides the ad-spend component (used
// for the stress test); otherwise the assumption's own ad-spend share is used.
func RemainingMargin(m MarginAssumptions, adSpendPct *float64) float64 {
	ad := m.AdSpendPct
	if adSpendPct != nil {
		ad = *adSpendPct
	}
	return 1 - ad - m.TeamPct - m.OverheadPct - m.TargetProfitPct
}

// SnapUpToLadder returns the first ladder rung greater than or equal to rawSP,
// or nil when the raw price exceeds the top of the ladder. The ladder is assumed
// ascending.
func SnapUpToLadder(rawSP float64, ladder []int) *int {
	for _, price := range ladder {
		if float64(price) >= rawSP {
			p := price
			return &p
		}
	}
	return nil
}

// GenerateSellingPrice reverses the unit economics into a margin-safe selling
// price snapped up to the approved ladder, then stress-tests it.
//
// ladder may be nil, in which case the brand's built-in ladder is used.
// stressAdSpendPct is the ad-spend share for the stress test (the API always
// passes StressAdSpendPct = 0.40). It mirrors generate_selling_price in
// app/pricing/selling_price.py.
func GenerateSellingPrice(
	designCP float64,
	fixed FixedVariableCosts,
	brand Brand,
	margins MarginAssumptions,
	ladder []int,
	stressAdSpendPct float64,
) (SellingPriceResult, error) {
	marginNormal := RemainingMargin(margins, nil)
	if marginNormal <= 0 {
		return SellingPriceResult{}, fmt.Errorf(
			"margin assumptions leave no room for cost; normal margin is not positive",
		)
	}

	otherFixed := fixed.TotalExcludingCP()
	fixedVariable := roundHalfEven(designCP+otherFixed, 2)
	rawSP := roundHalfEven(fixedVariable/marginNormal, 2)

	if ladder == nil {
		ladder = LadderFor(brand)
	}
	recommended := SnapUpToLadder(rawSP, ladder)

	var cpPct *float64
	if recommended != nil {
		v := roundHalfEven(designCP/float64(*recommended), 4)
		cpPct = &v
	}

	marginStress := RemainingMargin(margins, &stressAdSpendPct)
	// Compare the ladder integer against the UNROUNDED break-even under stress,
	// exactly as the Python engine does.
	survivesStress := recommended != nil && marginStress > 0 &&
		float64(*recommended) >= fixedVariable/marginStress

	warnings := rungWarnings(designCP, otherFixed, marginNormal, ladder, recommended)
	if recommended == nil {
		warnings = append(warnings,
			"Raw price exceeds the top of the ladder — review costs or category.")
	}

	return SellingPriceResult{
		RawSP:              rawSP,
		RecommendedSP:      recommended,
		CPPctAtRecommended: cpPct,
		PassesNormal:       recommended != nil,
		SurvivesStress:     survivesStress,
		Warnings:           warnings,
	}, nil
}

// rungWarnings explains, for every rung cheaper than the recommended one, the
// Design CP that rung would need. The strings are reproduced verbatim from the
// Python engine, including the ₹ (U+20B9) and ≤ (U+2264) glyphs, because the
// frontend surfaces them directly. When recommended is nil the loop never breaks
// and every rung gets a warning.
func rungWarnings(designCP, otherFixed, marginNormal float64, ladder []int, recommended *int) []string {
	warnings := []string{}
	for _, price := range ladder {
		if recommended != nil && price >= *recommended {
			break
		}
		maxCP := roundHalfEven(float64(price)*marginNormal-otherFixed, 2)
		if maxCP >= 0 {
			warnings = append(warnings, fmt.Sprintf(
				"₹%d needs Design CP ≤ ₹%s (currently ₹%s).",
				price, formatFixed(maxCP, 0), formatFixed(designCP, 0),
			))
		} else {
			warnings = append(warnings, fmt.Sprintf(
				"₹%d is not viable at these fixed costs.", price,
			))
		}
	}
	return warnings
}
